internal/documentmap/documentmap.go

```diff
@@ -200,11 +196,12 @@ func (m *Map) TotalDocsBytes() types.ByteCount {
 
 // called in tests as well as internally
 func (m *Map) getMapKey(doc *bson.Raw) MapKey {
-	key := MapKey("")
+	var keyBuffer bytes.Buffer
 	for _, keyName := range m.fieldNames {
 		value := doc.Lookup(keyName)
-		key += MapKey(value.Type) + MapKey(value.Value)
+		keyBuffer.WriteByte(byte(value.Type))
+		keyBuffer.Write(value.Value)
```

Comment:
How does this save memory usage?

Reply:
I'm not sure but I assume that the old code constructs new strings for each iteration, which is less efficient than continuously appending substrings to a buffer.

Reply:
> less efficient than continuously appending substrings to a buffer

I’m not sure of this. Consider libc’s `realloc` function, which solves the same problem of “extending” a given buffer. Note that it returns a new pointer, which is because there may not be enough room to extend the buffer, in which case a new buffer is allocated and the existing buffer’s contents are copied over.

Extending `keyBuffer` twice here may actually be less efficient than the status quo of extending `key` only once.

Reply:
My understanding is that the status quo does two string concatenations. So it's not just extending `key` once, but it's allocating two new strings each time.

I added `keyBuffer.Grow` to pre-allocate space on the buffer.

Reply:
Status quo allocates two small strings and concatenates them. That’s a smaller haul, I think, than two resizes of a large buffer.

Grow() is probably the way to go here.